File: src/lib/grading/engine.ts
import type { Channel, GradingProgress, Rubric, RubricCriterion, ScoreRecord, Trace } from '../../types';
import type { LLMClient } from '../llm/client';
import { GRADING_SCHEMA, buildProductPrompt, buildProductSystem, buildTracePrompt, buildTraceSystem } from './prompts';
import { aggregatePasses, type PassResult } from './aggregate';

const PASSES_PER_CRITERION = 3; // spec §5.3: ≥3 passes, report median + spread
const CONCURRENCY = 6;

interface RawPass {
  evidence?: { turnId?: number | null; quote: string; reasoning: string }[];
  anchorMatched?: string;
  score: number | string;
  selfConfidence?: string;
}

function normalizePass(raw: unknown, channel: Channel, source: { essay?: string; trace?: Trace }): PassResult {
  const r = raw as RawPass;
  let score: number | 'no-evidence';
  if (r.score === 'no-evidence' || r.score === undefined || r.score === null) {
    score = 'no-evidence';
  } else {
    const n = typeof r.score === 'number' ? r.score : parseInt(String(r.score), 10);
    score = Number.isFinite(n) ? Math.max(0, Math.min(5, Math.round(n))) : 'no-evidence';
  }

  // Verify quotes actually appear in the source (evidence-provenance guard, spec §4:
  // "No score without evidence"). Drop fabricated quotes; if all quotes for a scored
  // pass are fabricated, demote the pass to no-evidence. On the trace channel, quotes
  // must come from STUDENT turns specifically — this is the attribution guard's
  // client-side backstop: a quote of assistant text fails the lookup even if the
  // model claimed a student turnId for it.
  const normalize = (s: string) => s.replace(/\s+/g, ' ').replace(/["'‘’“”]/g, "'").toLowerCase();
  const studentTurns = (source.trace?.turns ?? []).filter((t) => t.speaker === 'student');

  const locateInStudentTurns = (quote: string): number | undefined => {
    const q = normalize(quote);
    return studentTurns.find((t) => normalize(t.text).includes(q))?.turnId;
  };

  const evidence = (r.evidence ?? [])
    .map((e) => {
      if (!e.quote) return null;
      if (channel === 'product') {
        return normalize(source.essay ?? '').includes(normalize(e.quote))
          ? { quote: e.quote, reasoning: e.reasoning ?? '' }
          : null;
      }
      // Trace: find the student turn the quote actually lives in; correct a wrong
      // turnId rather than trusting the model's citation.
      const actualTurnId = locateInStudentTurns(e.quote);
      return actualTurnId !== undefined ? { turnId: actualTurnId, quote: e.quote, reasoning: e.reasoning ?? '' } : null;
    })
    .filter((e): e is NonNullable<typeof e> => e !== null);

  if (score !== 'no-evidence' && (r.evidence ?? []).length > 0 && evidence.length === 0) {
    score = 'no-evidence';
  }

  return {
    score,
    selfConfidence: r.selfConfidence === 'high' || r.selfConfidence === 'low' ? r.selfConfidence : 'med',
    evidence,
    anchorMatched: r.anchorMatched,
  };
}

async function gradeCriterion(
  llm: LLMClient,
  criterion: RubricCriterion,
  channel: Channel,
  rubric: Rubric,
  source: { essay?: string; trace?: Trace },
): Promise<ScoreRecord> {
  const system = channel === 'product' ? buildProductSystem() : buildTraceSystem();
  const prompt =
    channel === 'product'
      ? buildProductPrompt(criterion, source.essay ?? '', rubric)
      : buildTracePrompt(criterion, source.trace as Trace, rubric);

  const passes: PassResult[] = [];
  for (let i = 0; i < PASSES_PER_CRITERION; i++) {
    // One criterion per call (spec §5.1); passes run sequentially per criterion so a
    // transient failure can be retried once without burning the whole batch.
    const request = {
      system,
      prompt,
      schema: GRADING_SCHEMA,
      useAdvisoryModel: criterion.referenceability === 'weak',
    };
    let raw: unknown;
    try {
      raw = await llm.completeJSON(request);
    } catch {
      // one retry per pass (transient API/parse failures); a second failure propagates
      raw = await llm.completeJSON(request);
    }
    passes.push(normalizePass(raw, channel, source));
  }

  return aggregatePasses({
    criterionId: criterion.criterionId,
    channel,
    referenceability: criterion.referenceability,
    passes,
    rubricVersion: rubric.version,
  });
}

/** Grade every criterion on both channels. Streams results via onResult as each
 *  criterion×channel completes (progressive UI, spec §10 cost/latency budget). */
export async function gradeSession(args: {
  llm: LLMClient;
  rubric: Rubric;
  essay: string;
  trace: Trace;
  onProgress?: (p: GradingProgress) => void;
  onResult?: (r: ScoreRecord) => void;
}): Promise<ScoreRecord[]> {
  const { llm, rubric, essay, trace, onProgress, onResult } = args;
  const jobs: { criterion: RubricCriterion; channel: Channel }[] = [];
  for (const c of rubric.criteria) {
    jobs.push({ criterion: c, channel: 'product' });
    jobs.push({ criterion: c, channel: 'trace' });
  }

  const results: ScoreRecord[] = [];
  let done = 0;
  let next = 0;

  async function worker() {
    while (next < jobs.length) {
      const job = jobs[next++];
      const record = await gradeCriterion(llm, job.criterion, job.channel, rubric, { essay, trace });
      results.push(record);
      done++;
      onProgress?.({ done, total: jobs.length, label: `${job.criterion.criterionId} · ${job.channel}` });
      onResult?.(record);
    }
  }

  await Promise.all(Array.from({ length: Math.min(CONCURRENCY, jobs.length) }, worker));
  return results;
}

File: backend/app/services/grading/engine.py
"""Mode A grading engine — port of TGFWA src/lib/grading/engine.ts.

Methodology (TGFWA spec §5): one criterion per LLM call (halo prevention),
both channels (product = essay, trace = dialogue), 3 sequential passes per
criterion with one retry each, evidence-provenance guard on every pass,
median + spread aggregation, concurrency 6 across criterion×channel jobs.

The LLM is injected as a callable `llm_json(system, prompt) -> dict` so tests
can drive the engine with a FakeLLM and the API layer wires in the configured
provider (core.llm.llm_chat_json + _extract_json).
"""

import re
from concurrent.futures import ThreadPoolExecutor, as_completed

from .aggregate import aggregate_passes
from .prompts import (build_product_prompt, build_product_system,
                      build_trace_prompt, build_trace_system)

PASSES_PER_CRITERION = 3  # spec §5.3: ≥3 passes, report median + spread
CONCURRENCY = 6


def _normalize_text(s: str) -> str:
    """Whitespace-collapse + quote-unify + lowercase, mirroring the TS guard."""
    s = re.sub(r"\s+", " ", s)
    s = re.sub(r"[\"'‘’“”]", "'", s)
    return s.lower()


def normalize_pass(raw: dict, channel: str, source: dict) -> dict:
    """Validate one raw LLM pass. Evidence-provenance guard (spec §4: "no score
    without evidence"): drop fabricated quotes; if all quotes for a scored pass
    are fabricated, demote the pass to no-evidence. On the trace channel quotes
    must come from STUDENT turns specifically — the attribution guard's
    server-side backstop: a quote of assistant text fails the lookup even if
    the model claimed a student turnId for it.
    """
    raw = raw if isinstance(raw, dict) else {}
    score = raw.get("score")
    if score in ("no-evidence", None):
        score = "no-evidence"
    else:
        try:
            n = int(round(float(score)))
            score = max(0, min(5, n))
        except (TypeError, ValueError):
            score = "no-evidence"

    student_turns = [t for t in (source.get("trace") or {}).get("turns", [])
                     if t.get("speaker") == "student"]

    def locate_in_student_turns(quote: str):
        q = _normalize_text(quote)
        for t in student_turns:
            if q in _normalize_text(t.get("text", "")):
                return t.get("turnId")
        return None

    raw_evidence = raw.get("evidence") or []
    if not isinstance(raw_evidence, list):
        raw_evidence = []
    evidence = []
    for e in raw_evidence:
        if not isinstance(e, dict) or not e.get("quote"):
            continue
        if channel == "product":
            if _normalize_text(e["quote"]) in _normalize_text(source.get("essay") or ""):
                evidence.append({"quote": e["quote"], "reasoning": e.get("reasoning", "")})
        else:
            # Trace: find the student turn the quote actually lives in; correct a
            # wrong turnId rather than trusting the model's citation.
            actual_turn_id = locate_in_student_turns(e["quote"])
            if actual_turn_id is not None:
                evidence.append({"turnId": actual_turn_id, "quote": e["quote"],
                                 "reasoning": e.get("reasoning", "")})

    if score != "no-evidence" and raw_evidence and not evidence:
        score = "no-evidence"

    self_conf = raw.get("selfConfidence")
    if self_conf not in ("high", "low"):
        self_conf = "med"

    return {
        "score": score,
        "selfConfidence": self_conf,
        "evidence": evidence,
        "anchorMatched": raw.get("anchorMatched"),
    }


def grade_criterion(llm_json, criterion: dict, channel: str, rubric: dict,
                    source: dict) -> dict:
    if channel == "product":
        system = build_product_system()
        prompt = build_product_prompt(criterion, source.get("essay") or "", rubric)
    else:
        system = build_trace_system()
        prompt = build_trace_prompt(criterion, source.get("trace") or {}, rubric)

    passes = []
    for _ in range(PASSES_PER_CRITERION):
        # One criterion per call; passes run sequentially per criterion so a
        # transient failure can be retried once without burning the whole batch.
        try:
            raw = llm_json(system, prompt)
        except Exception:
            raw = llm_json(system, prompt)  # one retry per pass; second failure propagates
        passes.append(normalize_pass(raw, channel, source))

    return aggregate_passes(
        criterion_id=criterion["criterionId"],
        channel=channel,
        referenceability=criterion.get("referenceability", "strong"),
        passes=passes,
        rubric_version=rubric.get("version", ""),
    )


def grade_session(*, llm_json, rubric: dict, essay: str, trace: dict,
                  on_progress=None, on_result=None) -> list:
    """Grade every criterion on both channels. Streams results via on_result as
    each criterion×channel completes (progressive persistence + SSE)."""
    jobs = [(c, channel)
            for c in rubric.get("criteria", [])
            for channel in ("product", "trace")]
    source = {"essay": essay, "trace": trace}

    results = []
    done = 0

    def run(job):
        criterion, channel = job
        return grade_criterion(llm_json, criterion, channel, rubric, source)

    with ThreadPoolExecutor(max_workers=min(CONCURRENCY, max(1, len(jobs)))) as pool:
        futures = {pool.submit(run, job): job for job in jobs}
        for future in as_completed(futures):
            criterion, channel = futures[future]
            record = future.result()  # propagate the first failure
            results.append(record)
            done += 1
            if on_progress:
                on_progress(done, len(jobs), f"{criterion['criterionId']} · {channel}")
            if on_result:
                on_result(record)
    return results
